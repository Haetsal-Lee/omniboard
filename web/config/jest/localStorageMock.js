class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }

  getAllItems() {
    return this.store;
  }
}

export {LocalStorageMock};