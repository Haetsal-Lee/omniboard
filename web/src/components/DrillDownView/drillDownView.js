import PropTypes from 'prop-types'
import React, { Component } from 'react';
import axios from 'axios';
import ReactJson from 'react-json-view'
import './drillDownView.scss'
import { Nav, NavItem } from 'react-bootstrap';
import { on, off } from 'dom-helpers/events';
import { capitalize } from '../Helpers/utils';
import LocalStorageMixin from 'react-localstorage';
import reactMixin from 'react-mixin';
import Plot from 'react-plotly.js';
import { ProgressWrapper } from '../Helpers/hoc';
import { parseServerError } from '../Helpers/utils';
import { toast } from 'react-toastify';
import { xAxisValues, DRILLDOWN_VIEW, scaleValues, SCALE_VALUE } from '../../constants/drillDownView.constants';

class JsonView extends React.PureComponent {
  static propTypes = {
    data: PropTypes.object
  };

  render() {
    const {data} = this.props;
    return (<ReactJson src={data} iconStyle="circle" collapsed={1} collapseStringsAfterLength={40}
                       enableClipboard={false} displayDataTypes={false} />);
  }
}

class DrillDownView extends Component {
  static propTypes = {
    height: PropTypes.number,
    runId: PropTypes.number,
    width: PropTypes.number
  };

  // Filter out state objects that need to be synchronized with local storage
  static defaultProps = {
    stateFilterKeys: ['selectedNavTab']
  };

  constructor(props) {
    super(props);
    this.state = {
      selectedNavTab: DRILLDOWN_VIEW.METRICS,
      isTableLoading: false,
      runsResponse: null,
      metricsResponse: null,
      selectedMetricNames: new Set(),
      selectedXAxis: xAxisValues[0],
      selectedYAxis: scaleValues[0]
    };
  }

  _handleSelectNavPill = (selectedKey) => {
    this.setState({
      selectedNavTab: selectedKey
    });
  };

  _stopWheel = event => {
    const {scrollTop, scrollHeight, clientHeight} = this.scrollableDiv;
    const maxScrollTop = scrollHeight - clientHeight;
    const scrollingUp = event.deltaY < 0;
    const scrollingDown = event.deltaY > 0;
    // Handle event only if scrolling up or down
    if (scrollingUp || scrollingDown) {
      // Stop event propagation to allow DrillDownView content to scroll
      if ((scrollTop > 0 && (scrollTop < maxScrollTop || scrollingUp)) || (scrollTop === 0 && maxScrollTop > 0 && scrollingDown)) {
        event.stopPropagation();
      }
    }
    if (this.state.selectedNavTab === DRILLDOWN_VIEW.CAPTURED_OUT) {
      // Stop event propagation for horizontal scroll to allow CAPTURED_OUT content to be scrolled
      if (event.deltaX !== 0) {
        event.stopPropagation();
      }
    }
  };

  componentWillMount() {
    const {runId} = this.props;
    if (runId) {
      this.setState({
        isTableLoading: true
      });
      axios.all([
        axios.get(`/api/v1/Runs/${runId}`, {
          params: {
            select: 'captured_out,info,meta,host,experiment'
          }
        }),
        axios.get('/api/v1/Metrics', {
          params: {
            query: `{"run_id":"${runId}"}`
          }
        })
      ])
      .then(axios.spread((runsResponse, metricsResponse) => {
        const runsResponseData = runsResponse.data;
        const metricsResponseData = metricsResponse.data;
        this.setState({
          isTableLoading: false,
          runsResponse: runsResponseData,
          metricsResponse: metricsResponseData
        });
      }))
      .catch(error => {
        this.setState({isTableLoading: false});
        toast.error(parseServerError(error));
      });
    }
  }

  componentDidMount() {
    on(this.scrollableDiv, 'wheel', this._stopWheel);
  }

  componentWillUnmount() {
    off(this.scrollableDiv, 'wheel', this._stopWheel);
  }

  _handleXAxisChange = (event) => {
    this.setState({
      selectedXAxis: event.target.value
    });
  };

  _handleYAxisChange = (event) => {
    this.setState({
      selectedYAxis: event.target.value
    });
  };

  _handleMetricNamesChange = (event) => {
    const {selectedMetricNames} = this.state;
    const value = event.target.value;
    if (event.target.checked) {
      selectedMetricNames.add(value);
    } else {
      selectedMetricNames.delete(value);
    }
    this.setState({
      selectedMetricNames
    })
  };

  getMetricsPlot() {
    const {metricsResponse, selectedMetricNames, selectedXAxis, selectedYAxis} = this.state;
    let metricsResponseMap = {},
      metricNames = [];
    if (metricsResponse && metricsResponse.length) {
      metricsResponseMap = metricsResponse.reduce((map = {}, metric) => {
        map[metric.name] = metric;
        return map;
      }, {});
      metricNames = Object.keys(metricsResponseMap);
    }
    if (!metricNames.length) {
      return (
        <div className="alert alert-warning">No metrics are available to plot</div>
      )
    }
    const plotData = Array.from(selectedMetricNames).map(metricName => {
      return {
        type: 'scatter',
        mode: 'lines+points',
        name: metricName,
        x: metricsResponseMap[metricName][selectedXAxis],
        y: metricsResponseMap[metricName]['values']
      };
    });
    const yAxisLayout = selectedYAxis === SCALE_VALUE.LOGARITHMIC ? {type: 'log', autorange: true} : {};
    return (
      <div>
        <div className="metrics-plot-left">
          <h4>Metrics to plot</h4>
          <div id="plot-metric-names">
            {metricNames.map( (metricName, i) => {
              return (
                <div key={i} className="checkbox">
                  <label>
                    <input test-attr={"plot-metric-name-" + i} type="checkbox" value={metricName} checked={selectedMetricNames.has(metricName)} onChange={this._handleMetricNamesChange}/>
                    {metricName}
                  </label>
                </div>
              )
            })}
          </div>
          <h4>X-Axis Type</h4>
          <div id="plot-x-axis-types">
            {xAxisValues.map( (value, i) => {
              return (
                <div key={i} className="radio">
                  <label>
                    <input test-attr={"plot-x-axis-" + i} type="radio" name="x-axis" value={value} checked={selectedXAxis === value} onChange={this._handleXAxisChange} />
                    {capitalize(value)}
                  </label>
                </div>
              )
            })}
          </div>
          <h4>Y-Axis Type</h4>
          <div id="plot-y-axis-types">
            {scaleValues.map( (value, i) => {
              return (
                <div key={i} className="radio">
                  <label>
                    <input test-attr={"plot-y-axis-" + i} type="radio" name="y-axis" value={value} checked={selectedYAxis === value} onChange={this._handleYAxisChange} />
                    {capitalize(value)}
                  </label>
                </div>
              )
            })}
          </div>
        </div>
        <div className="metrics-plot-content">
          <Plot
            data={plotData}
            layout={{
              width: 700,
              height: 300,
              title: 'Metrics Plot',
              yaxis: yAxisLayout
            }}
          />
        </div>
      </div>
    );
  }

  render () {
    const {selectedNavTab, isTableLoading, runsResponse} = this.state;
    const {width, height, runId} = this.props;
    const experimentName = runsResponse !== null && 'experiment' in runsResponse ? runsResponse.experiment.name : '';
    // Adjust the width of scrollbar from total width
    const style = {
      height: height,
      width: width - 17,
    };
    const getTabContent = (key) => {
      let content = '';
      if (runsResponse) {
        switch (key) {
          case DRILLDOWN_VIEW.EXPERIMENT:
            content = <div id={DRILLDOWN_VIEW.EXPERIMENT}><JsonView data={runsResponse.experiment}/></div>;
            break;
          case DRILLDOWN_VIEW.HOST_INFO:
            content = <div id={DRILLDOWN_VIEW.HOST_INFO}><JsonView data={runsResponse.host}/></div>;
            break;
          case DRILLDOWN_VIEW.RUN_INFO:
            content = <div id={DRILLDOWN_VIEW.RUN_INFO}><JsonView data={runsResponse.info}/></div>;
            break;
          case DRILLDOWN_VIEW.CAPTURED_OUT:
            content = <div id={DRILLDOWN_VIEW.CAPTURED_OUT}><pre>{runsResponse.captured_out}</pre></div>;
            break;
          case DRILLDOWN_VIEW.META_INFO:
            content = <div id={DRILLDOWN_VIEW.META_INFO}><JsonView data={runsResponse.meta}/></div>;
            break;
          case DRILLDOWN_VIEW.METRICS:
            content = <div id={DRILLDOWN_VIEW.METRICS}>{this.getMetricsPlot()}</div>;
            break;
          default:
        }
      }
      return <ProgressWrapper loading={isTableLoading}>{content}</ProgressWrapper>;
    };
    return(
      <div style={style}>
        <nav className="navbar navbar-light">
          <span className="navbar-brand">Details for: <strong>{experimentName}</strong> (Id: {runId}) </span>
        </nav>
        <div className="sidebar-wrapper">
          <div className="sidebar-container full-height">
            <div id="sidebar">
              <Nav bsStyle="pills" stacked activeKey={selectedNavTab} onSelect={this._handleSelectNavPill}>
                <NavItem eventKey={DRILLDOWN_VIEW.METRICS}>
                  Metrics Plot
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.CAPTURED_OUT}>
                  Captured Out
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.EXPERIMENT}>
                  Experiment Details
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.HOST_INFO}>
                  Host Info
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.RUN_INFO}>
                  Run Info
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.META_INFO}>
                  Meta Info
                </NavItem>
              </Nav>
            </div>
          </div>
          <div className="sidebar-content full-height">
            <div className="tab-content" ref={div => (this.scrollableDiv = div || this.scrollableDiv)}>
              {getTabContent(selectedNavTab)}
            </div>
          </div>
        </div>
      </div>
    )
  }
}

reactMixin(DrillDownView.prototype, LocalStorageMixin);
export {DrillDownView};
