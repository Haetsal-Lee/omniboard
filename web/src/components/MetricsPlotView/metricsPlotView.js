import React, {Component} from "react";
import { SCALE_VALUE, scaleValues, xAxisValues } from "../../constants/drillDownView.constants";
import { capitalize } from "../Helpers/utils";
import Plot from "react-plotly.js";
import PropTypes from 'prop-types'
import reactMixin from "react-mixin";
import LocalStorageMixin from "react-localstorage";

const DEFAULT_SELECTION_KEY = "MetricsPlotView|default";

class MetricsPlotView extends Component {
  static propTypes = {
    metricsResponse: PropTypes.array,
    runId: PropTypes.any
  };

  // Filter out state objects that need to be synchronized with local storage
  static defaultProps = {
    stateFilterKeys: ['selectedMetricNames', 'selectedXAxis', 'selectedYAxis']
  };

  constructor(props) {
    super(props);
    this.state = {
      selectedMetricNames: [],
      selectedXAxis: xAxisValues[0],
      selectedYAxis: scaleValues[0]
    };
  }

  _handleXAxisChange = (event) => {
    const value = event.target.value;
    this.setState({
      selectedXAxis: value
    });
    // update local storage
    this._updateDefaultSelection({selectedXAxis: value});
  };

  _handleYAxisChange = (event) => {
    const value = event.target.value;
    this.setState({
      selectedYAxis: value
    });
    // update local storage
    this._updateDefaultSelection({selectedYAxis: value});
  };

  _handleMetricNamesChange = (event) => {
    const {selectedMetricNames} = this.state;
    let newMetricNames = Object.assign([], selectedMetricNames);
    const value = event.target.value;
    if (event.target.checked) {
      if (!newMetricNames.includes(value)) {
        newMetricNames.push(value);
      }
    } else {
      newMetricNames = newMetricNames.filter(metricName => metricName !== value);
    }
    this.setState({
      selectedMetricNames: newMetricNames
    });
    // update local storage
    this._updateDefaultSelection({selectedMetricNames: newMetricNames});
  };

  /**
   * Update the default selection of metrics plot in local storage
   * @param selection
   * @private
   */
  _updateDefaultSelection = (selection) => {
    const value = Object.assign({}, JSON.parse(localStorage.getItem(DEFAULT_SELECTION_KEY)), selection);
    localStorage.setItem(DEFAULT_SELECTION_KEY, JSON.stringify(value));
  };

  _setDefaultSelection = () => {
    const defaultSelection = JSON.parse(localStorage.getItem(DEFAULT_SELECTION_KEY));
    if (defaultSelection) {
      const metricNames = this.props.metricsResponse.map(metric => metric.name);
      const selectedMetricNames = defaultSelection.selectedMetricNames ? defaultSelection.selectedMetricNames.filter(name => metricNames.includes(name)) : [];
      this.setState({
        selectedMetricNames,
        selectedXAxis: defaultSelection.selectedXAxis || '',
        selectedYAxis: defaultSelection.selectedYAxis || ''
      });
    }
  };

  componentDidMount() {
    this._setDefaultSelection();
  }

  render() {
    const {metricsResponse, runId} = this.props;
    const {selectedMetricNames, selectedXAxis, selectedYAxis} = this.state;
    let metricsResponseMap = {},
      metricNames = [];
    if (metricsResponse && metricsResponse.length) {
      metricsResponseMap = metricsResponse.reduce((map = {}, metric) => {
        map[metric.name] = metric;
        return map;
      }, {});
      metricNames = Object.keys(metricsResponseMap);
    }
    if (!metricNames.length) {
      return (
        <div className="alert alert-warning">No metrics are available to plot</div>
      )
    }
    const plotData = Array.from(selectedMetricNames).map(metricName => {
      return {
        type: 'scatter',
        mode: 'lines+points',
        name: metricName,
        x: metricsResponseMap[metricName][selectedXAxis],
        y: metricsResponseMap[metricName]['values']
      };
    });
    const yAxisLayout = selectedYAxis === SCALE_VALUE.LOGARITHMIC ? {type: 'log', autorange: true} : {};
    return (
      <div>
        <div className="metrics-plot-left">
          <h4>Metrics to plot</h4>
          <div id="plot-metric-names">
            {metricNames.map( (metricName, i) => {
              return (
                <div key={"MetricName" + runId + i} className="checkbox">
                  <label>
                    <input test-attr={"plot-metric-name-" + i} type="checkbox" value={metricName} checked={selectedMetricNames.includes(metricName)} onChange={this._handleMetricNamesChange}/>
                    {metricName}
                  </label>
                </div>
              )
            })}
          </div>
          <h4>X-Axis Type</h4>
          <div id="plot-x-axis-types">
            {xAxisValues.map( (value, i) => {
              return (
                <div key={"XAxisPlot" + runId + i} className="radio">
                  <label>
                    <input test-attr={"plot-x-axis-" + i} key={"XAxisPlotInput" + runId + i} type="radio" value={value} checked={selectedXAxis === value} onChange={this._handleXAxisChange} />
                    {capitalize(value)}
                  </label>
                </div>
              )
            })}
          </div>
          <h4>Y-Axis Type</h4>
          <div id="plot-y-axis-types">
            {scaleValues.map( (value, i) => {
              return (
                <div key={"YAxisPlot" + runId + i} className="radio">
                  <label>
                    <input test-attr={"plot-y-axis-" + i} key={"YAxisPlotInput" + runId + i} type="radio" value={value} checked={selectedYAxis === value} onChange={this._handleYAxisChange} />
                    {capitalize(value)}
                  </label>
                </div>
              )
            })}
          </div>
        </div>
        <div className="metrics-plot-content">
          <Plot
            data={plotData}
            layout={{
              width: 700,
              height: 300,
              title: 'Metrics Plot',
              yaxis: yAxisLayout
            }}
          />
        </div>
      </div>
    );
  }
}

reactMixin(MetricsPlotView.prototype, LocalStorageMixin);
export {MetricsPlotView};
