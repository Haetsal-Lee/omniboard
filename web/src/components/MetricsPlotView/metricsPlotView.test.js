import React from 'react';
import { MetricsPlotView } from './metricsPlotView';
import mockAxios from 'jest-mock-axios';
import { X_AXIS_VALUE, SCALE_VALUE } from '../../constants/drillDownView.constants';
import { LocalStorageMock } from '../../../config/jest/localStorageMock';

describe('MetricsPlotView', () => {
  let wrapper = null;

  const metricsResponseData = [
    {"_id":"5a2b5a8c9c7a505a652f6127","name":"pretrain.train.loss","run_id":222, "steps":[0,1,2,3,4],
      "values":[0.7159541824544438,0.3840367944955761,0.3469185283233073,0.30483262065173106,0.28915774130337507],
      "timestamps":["2017-12-09T03:37:44.425Z","2017-12-09T03:41:54.414Z","2017-12-09T03:46:01.766Z","2017-12-09T03:50:07.365Z","2017-12-09T03:54:12.560Z"]},
    {"_id":"5a2b5aa09c7a505a652f6146","name":"pretrain.val.loss","run_id":223, "steps":[0,1,2,3,4],
      "values":[0.32177006650114165,0.23237958704995795,0.23340759051386187,0.21925230575196739,0.20541178824900605],
      "timestamps":["2017-12-09T03:38:01.945Z","2017-12-09T03:42:11.673Z","2017-12-09T03:46:18.843Z","2017-12-09T03:50:24.377Z","2017-12-09T03:54:29.752Z"]}];

  beforeEach(() => {
    wrapper = mount(
      <MetricsPlotView metricsResponse={metricsResponseData} runId={222} localStorageKey={"metricsPlot|222"}/>
    );
  });

  afterEach(() => {
    // cleaning up the mess left behind the previous test
    mockAxios.reset();
    jest.clearAllMocks();
  });

  it('should render correctly', async () => {

    expect(wrapper).toMatchSnapshot();
  });

  it('should set default selection correctly', () => {
    /* eslint-disable no-global-assign */
    localStorage = {getItem: () => '{"selectedMetricNames": ["pretrain.val.loss", "invalid"], "selectedXAxis": "time", "selectedYAxis": "linear"}'};
    wrapper.instance()._setDefaultSelection();

    expect(wrapper.state().selectedMetricNames).toHaveLength(1);
    expect(wrapper.state().selectedXAxis).toEqual('time');
    expect(wrapper.state().selectedYAxis).toEqual('linear');

    localStorage = {getItem: () => '{}'};
    wrapper.instance()._setDefaultSelection();

    expect(wrapper.state().selectedMetricNames).toHaveLength(0);
    expect(wrapper.state().selectedXAxis).toEqual('');
    expect(wrapper.state().selectedYAxis).toEqual('');
    // reset localStorage
    localStorage = new LocalStorageMock;
  });

  it('should show alert when no metrics are available', () => {
    wrapper = mount(
      <MetricsPlotView metricsResponse={[]} runId={222} localStorageKey={"metricsPlot|222"}/>
    );

    expect(wrapper.find('.alert')).toHaveLength(1);
  });

  describe('should handle', () => {
    it('metricNamesChange correctly', async () => {

      expect(wrapper.find('[test-attr="plot-metric-name-0"]')).toHaveLength(1);
      wrapper.find('[test-attr="plot-metric-name-0"]').simulate('change', {
        target: {
          value: 'pretrain.train.loss',
          checked: true
        }
      });
      wrapper.find('[test-attr="plot-metric-name-0"]').simulate('change', {
        target: {
          value: 'pretrain.train.loss',
          checked: true
        }
      });

      expect(wrapper.state().selectedMetricNames).toHaveLength(1);

      wrapper.find('[test-attr="plot-metric-name-0"]').simulate('change', {
        target: {
          value: 'pretrain.train.loss',
          checked: false
        }
      });

      expect(wrapper.state().selectedMetricNames).toHaveLength(0);
    });

    it('x-axis change correctly', async () => {
      wrapper.find('[test-attr="plot-x-axis-0"]').simulate('change', {target: {value: X_AXIS_VALUE.TIME}});

      expect(wrapper.state().selectedXAxis).toEqual(X_AXIS_VALUE.TIME);
    });

    it('y-axis change correctly', async () => {
      wrapper.find('[test-attr="plot-y-axis-0"]').simulate('change', {target: {value: SCALE_VALUE.LOGARITHMIC}});

      expect(wrapper.state().selectedYAxis).toEqual(SCALE_VALUE.LOGARITHMIC);
    });
  });
});
