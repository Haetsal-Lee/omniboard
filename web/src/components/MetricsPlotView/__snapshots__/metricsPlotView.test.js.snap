// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricsPlotView should render correctly 1`] = `
<MetricsPlotView
  localStorageKey="metricsPlot|222"
  metricsResponse={
    Array [
      Object {
        "_id": "5a2b5a8c9c7a505a652f6127",
        "name": "pretrain.train.loss",
        "run_id": 222,
        "steps": Array [
          0,
          1,
          2,
          3,
          4,
        ],
        "timestamps": Array [
          "2017-12-09T03:37:44.425Z",
          "2017-12-09T03:41:54.414Z",
          "2017-12-09T03:46:01.766Z",
          "2017-12-09T03:50:07.365Z",
          "2017-12-09T03:54:12.560Z",
        ],
        "values": Array [
          0.7159541824544438,
          0.3840367944955761,
          0.3469185283233073,
          0.30483262065173106,
          0.28915774130337507,
        ],
      },
      Object {
        "_id": "5a2b5aa09c7a505a652f6146",
        "name": "pretrain.val.loss",
        "run_id": 223,
        "steps": Array [
          0,
          1,
          2,
          3,
          4,
        ],
        "timestamps": Array [
          "2017-12-09T03:38:01.945Z",
          "2017-12-09T03:42:11.673Z",
          "2017-12-09T03:46:18.843Z",
          "2017-12-09T03:50:24.377Z",
          "2017-12-09T03:54:29.752Z",
        ],
        "values": Array [
          0.32177006650114165,
          0.23237958704995795,
          0.23340759051386187,
          0.21925230575196739,
          0.20541178824900605,
        ],
      },
    ]
  }
  runId={222}
  stateFilterKeys={
    Array [
      "selectedMetricNames",
      "selectedXAxis",
      "selectedYAxis",
    ]
  }
>
  <div>
    <div
      className="metrics-plot-left"
    >
      <h4>
        Metrics to plot
      </h4>
      <div
        id="plot-metric-names"
      >
        <div
          className="checkbox"
          key="MetricName2220"
        >
          <label>
            <input
              checked={false}
              onChange={[Function]}
              test-attr="plot-metric-name-0"
              type="checkbox"
              value="pretrain.train.loss"
            />
            pretrain.train.loss
          </label>
        </div>
        <div
          className="checkbox"
          key="MetricName2221"
        >
          <label>
            <input
              checked={false}
              onChange={[Function]}
              test-attr="plot-metric-name-1"
              type="checkbox"
              value="pretrain.val.loss"
            />
            pretrain.val.loss
          </label>
        </div>
      </div>
      <h4>
        X-Axis Type
      </h4>
      <div
        id="plot-x-axis-types"
      >
        <div
          className="radio"
          key="XAxisPlot2220"
        >
          <label>
            <input
              checked={true}
              key="XAxisPlotInput2220"
              onChange={[Function]}
              test-attr="plot-x-axis-0"
              type="radio"
              value="steps"
            />
            Steps
          </label>
        </div>
        <div
          className="radio"
          key="XAxisPlot2221"
        >
          <label>
            <input
              checked={false}
              key="XAxisPlotInput2221"
              onChange={[Function]}
              test-attr="plot-x-axis-1"
              type="radio"
              value="timestamps"
            />
            Timestamps
          </label>
        </div>
      </div>
      <h4>
        Y-Axis Type
      </h4>
      <div
        id="plot-y-axis-types"
      >
        <div
          className="radio"
          key="YAxisPlot2220"
        >
          <label>
            <input
              checked={true}
              key="YAxisPlotInput2220"
              onChange={[Function]}
              test-attr="plot-y-axis-0"
              type="radio"
              value="linear"
            />
            Linear
          </label>
        </div>
        <div
          className="radio"
          key="YAxisPlot2221"
        >
          <label>
            <input
              checked={false}
              key="YAxisPlotInput2221"
              onChange={[Function]}
              test-attr="plot-y-axis-1"
              type="radio"
              value="logarithmic"
            />
            Logarithmic
          </label>
        </div>
      </div>
    </div>
    <div
      className="metrics-plot-content"
    >
      <PlotlyComponent
        data={Array []}
        debug={false}
        fit={false}
        layout={
          Object {
            "height": 300,
            "title": "Metrics Plot",
            "width": 700,
            "xaxis": Object {
              "autorange": true,
              "range": Array [
                -1,
                6,
              ],
            },
            "yaxis": Object {
              "autorange": true,
              "range": Array [
                -1,
                4,
              ],
            },
          }
        }
        style={
          Object {
            "display": "inline-block",
            "position": "relative",
          }
        }
        useResizeHandler={false}
      >
        <div
          style={
            Object {
              "display": "inline-block",
              "position": "relative",
            }
          }
        />
      </PlotlyComponent>
    </div>
  </div>
</MetricsPlotView>
`;
